### ISSUE TITLE
The loss of variance_adaptor(Mandarin dataset)

### ISSUE THREAD

**Issue by humanlost** [ming024/FastSpeech2#1]
Hi, I used Mandarin dataset(BIAOBEI) to train FastSpeech2.
The loss of mel and PostNet mel seems no problem.
But I find out that the loss of variance_adaptor (Duration Loss, F0 Loss and Energy Loss) is really high.

The following is a part of my log:
Epoch [191/1000], Step [115650/608000]:
Total Loss: 68.7120, Mel Loss: 0.2892, Mel PostNet Loss: 0.2889, Duration Loss: 2.2572, F0 Loss: 59.7572, Energy Loss: 6.1195;
Time Used: 28493.331s, Estimated Time Remaining: 97800.871s.

How could I solve this?

Thank you.

**Comment by rishikksh20** [comment]
@humanlost @ming024 Error of Variance Adaptor should be high as the actual value of these component is very high but that's not the issue if training error and validation error is decreasing, you can normalize F0 and Energy for lower error. In paper the author mentioned that the well trained model has MAE (mean absolute error) between actual and predicted f0 is around 20 and same for energy is around 0.13.

**Comment by ming024** [comment]
@rishikksh20 That's true. But I still wondering that whether the performance will be better if the loss terms are properly weighted. Still need more experiments.

**Comment by humanlost** [comment]
@ming024 @rishikksh20  Thank you for your answers. 

Before I trained the Mandarin model, I used to download the FastSpeech2 pretrained model(checkpoint_300000.pth.tar). It really has good performance. But you said that this problem happens in your experiments on the LJSpeech dataset, so What is the final loss of your model?

The difference between training error and validation error makes the audio generated by my model during testing much worse than the audio generated during training.

**Comment by rishikksh20** [comment]
@ming024 weighted loss sounds good, but what I do that instead of using the embedding layer for energy and f0 I use linear dense layer `(torch.nn.Linear(256, 256))` for respectively, what I do is to convert f0 and energy value predicted to a one-hot vector of 256 and then passed respective one-hot vector to the respective linear dense layer of 256 and then add the output of dense layer f0 and energy component to the hidden vectors.

**Comment by rishikksh20** [comment]
Also instead of using a heavy waveglow model you can use fast and compact https://github.com/seungwonpark/melgan for mel 2 wav generation. As you have used Nvidia's preprocessing you don't have to change anything for melgan vocoder just add following code:
```
import torch
vocoder = torch.hub.load('seungwonpark/melgan', 'melgan')
vocoder.eval()
mel = torch.randn(1, 80, 234) # use your own mel-spectrogram here

if torch.cuda.is_available():
    vocoder = vocoder.cuda()
    mel = mel.cuda()

with torch.no_grad():
    audio = vocoder.inference(mel)
```
This Melgan implementation uses the same pre-processing (Nvidia's preprocessing) as used in this repo, so no modification required just to maintain the same Mel shape. 
The output of WaveGlow sounds like faded audio whereas the output of MelGAN sounds like amplified audio which is good to understand the voice quality.

**Comment by ming024** [comment]
@humanlost This is my loss on LJSpeech dataset at the 110000-th step of training.

Epoch [147/1000], Step [110000/752000]:
Total Loss: 25.9775, Mel Loss: 0.2172, Mel PostNet Loss: 0.2172, Duration Loss: 1.3491, F0 Loss: 21.1868, Energy Loss: 3.0071;  
Time Used: 40295.231s, Estimated Time Remaining: 189252.087s.

It seems that the F0 loss on your dataset is larger. Did you handle the tone symbols in Chinese language? I haven't tried this implementation on non-English datasets, but I think that is a big difference between Chinese and English.

**Comment by ming024** [comment]
@rishikksh20 I think using embedding layers and linear layer + one-hot encoding does not make any difference? And thank for your suggestion about using MelGAN instead of WaveGlow. I will tri it.

**Comment by humanlost** [comment]
Yes, I use a Chinese phoneme system. But F0 loss seems to be large.

I will continue to try to train the Chinese model and find the reason why Loss is too high.

Thank you.

**Comment by chazo1994** [comment]
@humanlost  Did you found the reason why your loss is too high, and how to reduce it?

**Comment by guomin** [comment]
@chazo1994 @humanlost 你们解决了吗，我遇到了同样的问题。奇怪的是，训练过程产生的sample音频效果还好，但测试合成的时候，听起来非常奇怪。
